server.go

```diff
@@ -235,45 +239,66 @@ func (s *Server) handleRead(pubKey credentials.StaticSizedPublicKey, done <-chan
 // the handler. The connection client's public key is injected into the context,
 // so the handler is able to identify the caller.
 func (s *Server) handleMessageRequest(pubKey credentials.StaticSizedPublicKey, r *message.Request) {
+	if err := s.validateMessageRequest(r); err != nil {
+		log.Printf("error validating request: %s", err)
+		return
+	}
+
 	methodName := r.GetMethod()
-	if md, ok := s.service.methods[methodName]; ok {
-		// Create a decoder function to unmarshal the message
-		dec := func(v interface{}) error {
-			return UnmarshalProtoMessage(r.GetPayload(), v)
-		}
+	md := s.service.methods[methodName]
+	// Create a decoder function to unmarshal the message
+	dec := func(v interface{}) error {
+		return UnmarshalProtoMessage(r.GetPayload(), v)
+	}
 
-		// Inject the peer's public key into the context so the handler can use it
-		ctx := peer.NewContext(context.Background(), &peer.Peer{PublicKey: pubKey})
-		v, herr := md.Handler(s.service.serviceImpl, ctx, dec)
+	// Inject the peer's public key into the context so the handler can use it
+	ctx := peer.NewContext(context.Background(), &peer.Peer{PublicKey: pubKey})
+	v, herr := md.Handler(s.service.serviceImpl, ctx, dec)
 
-		msg, err := message.NewResponse(r.GetCallId(), v, herr)
-		if err != nil {
-			return
-		}
+	msg, err := message.NewResponse(r.GetCallId(), v, herr)
+	if err != nil {
+		return
+	}
 
-		replyMsg, err := MarshalProtoMessage(msg)
-		if err != nil {
-			return
-		}
+	replyMsg, err := MarshalProtoMessage(msg)
+	if err != nil {
+		return
+	}
 
-		if err := s.sendMsg(ctx, pubKey, replyMsg); err != nil {
-			log.Printf("error sending message: %s", err)
-		}
+	if err := s.sendMsg(ctx, pubKey, replyMsg); err != nil {
+		log.Printf("error sending message: %s", err)
 	}
 }
 
 // handleMessageResponse finds the call which matches the method call id of the
 // response and sends the payload to the call channel.
-func (s *Server) handleMessageResponse(r *message.Response) {
+func (s *Server) handleMessageResponse(pubKey credentials.StaticSizedPublicKey, r *message.Response) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	callID := r.GetCallId()
-	if call, ok := s.methodCalls[callID]; ok {
-		call <- r
+	call, err := s.methodCalls.GetMessageResponseChannelForPublicKey(pubKey, callID)
+	if err != nil {
+		log.Printf("error handling message response: %s", err)
+		return
+	}
+
+	call <- r
+	s.removeMethodCall(pubKey, callID) // Delete the call now that we have completed the request/response cycle
+}
+
+func (s *Server) validateMessageRequest(r *message.Request) error {
+	methodName := r.GetMethod()
+	if _, ok := s.service.methods[methodName]; !ok {
+		return fmt.Errorf("invalid method: %v", methodName)
+	}
 
-		s.removeMethodCall(callID) // Delete the call now that we have completed the request/response cycle
+	callId := r.GetCallId()
+	if id, err := uuid.Parse(callId); err != nil || id.Version() != 4 {
+		return fmt.Errorf("invalid callId: %v", callId)
```

Comment:
Wrap the original error?
```suggestion
		return fmt.Errorf("invalid callId %s: %w", callId, err)
```